scuba/__main__.py

```diff
@@ -415,6 +416,12 @@ def get_docker_cmdline(self):
 
         args += self.options
 
+        # Precedence: (1) command-line -d, (2) alias docker_args, (3) top-level docker_args
+        if self.docker_args:
+            args += self.docker_args
+        elif self.context.docker_args is not None:
+            args += self.context.docker_args
+
```

Comment:
Given that other options (e.g. `image`) respect this same precedence, this makes sense. But I could also see an additive behavior being useful; IOW the resulting set of docker args would be the union of the three.

There is some precedent for "merging" in the alias `environment` which is [combined with](https://github.com/JonathonReinhart/scuba/blob/v2.7.0/scuba/config.py#L329-L331) the top-level `environment`.

Starting with just `.scuba.yml`, I was hoping there would be some YAML syntax which would allow for explicitly pulling the top-level `docker_args` into the per-alias ones, but a) it's a string, not a list, and b) [there is no mechanism for lists](https://stackoverflow.com/questions/9254178) anyway.

**Rationale:**
- At least in the way I use scuba, things in `.scuba.yml` are "permanent" while things on the command line are one-offs for testing, etc.  I can imagine use cases where I want to add additional docker args (e.g. another bind-mount), but I can't think of any docker args that I would specifically want to omit/override.
- If scuba were additive but one needed to override for a one-off, I feel like it would be easier to comment-out the line in `.scuba.yml`, rather than the other way around

Thoughts?

Reply:
I was on the fence whether to go with the additive or override approach. I think I'm on board with reworking it as additive after reading though your thoughts.

Reply:
Thinking about this more, I think it would be nice to support both behaviors. 

Consider this (contrived) example where scuba is used to run tests for some software that is designed with "recommended" and "minimum" system requirements, and it takes advantage of the `--memory` docker argument to enforce such limits during test runs:
 
 ```yaml
image: debian:buster
# By default, we want to run with 2GB of memory - our software's "recommended" system requirements.
docker_args: --memory 2G
 
aliases:
  test-quick:
    script: ./run_quick_tests.sh

  test-all:
    script: ./run_all_tests.sh

  # Run tests in a container with limited memory to make sure we support low-end hardware
  test-mem-limited:
    docker_args: --memory 512M
    script: ./run_all_tests.sh
```

With the merging strategy, this use case of scuba would need to specify their default `--memory` argument on every alias , even if most aliases just wanted the same configuration. Trivial in this example, but easy to imagine this blowing up if there were many more aliases, and if the default args were significantly more complex (maybe they're also tweaking other hardware configs like CPU, read/write speeds, etc).

 A possible solution would be to support both a "simple" and "extended" `docker_args` definition. An alias could specify `docker_args` as a string, where it would be used directly, or as a dict with additional options:
 
 ```yaml
 image: debian:buster
 docker_args: --memory 2G
 
aliases:
  cmd1:
    # This alias ignores the global docker_args
    docker_args:
        strategy: override
        value: --memory 512M
  cmd2:
    # This alias uses the global docker_args
    docker_args:
        strategy: merge  # the default behavior - optional to specify here
        value: -p 80:80
  cmd3:
    # This alias also uses the global docker_args
    docker_args: -p 80:80 -p 443:443
 ```
 
With all the said, I think the interaction of args specified on the scuba CLI (`-d`) can only behave one way, and merging seems like the more reasonable default behavior.
 
 Let me know what you think on this.

Reply:
That's a great counterexample, and  a workable solution.

Here's another idea, inspired by something I saw elsewhere: We could use a `+` suffix on the key to denote an append behavior, and use replacement otherwise.

```yaml
docker_args: --memory 2G

aliases:
  test:
    docker_args: --memory 512M
  test2:
    docker_args+: --privileged
```

The two would be mutually exclusive.

It's a little less verbose than yours, and hopefully equivalently powerful.

Similarly, we could use e.g., `-d` and `-D` command line args.

Another random thought: we could support both shlex-split strings and lists, if we wanted to in the future.

Reply:
If I may drop my 2 cents here: I think having the ability to override the parent docker args is important for this feature, but I think it is the corner case. I expect that a majority of users are going to use/expect the appending behavior as opposed to the overriding behavior. To that end, perhaps it would make more sense to have the `docker_args` key (the one that is most logical and easy to remember) use the append behavior and use a key name of `docker_args-` or similar to use the override behavior.

Reply:
I don't think I am really a fan of inventing new syntax like a `+` or `-` character and strapping it on top of YAML, especially if it's only used for one specific key in all of scuba. 

The scuba file should prioritize readability, even if you're not familiar with scuba, over how fast you can type it. It is a configuration file, not a CLI command you have to type frequently.

scuba already has one special YAML construct: `!from_yaml`. Maybe sharing syntax with that would be best, and good middle ground between an overly verbose syntax and quick to write:

```yaml
image: debian:buster
docker_args: --memory 2G

aliases:
 cmd1:
   # This alias ignores the global docker_args
   docker_args: !override --memory 512M
 cmd2:
   # This alias uses the global docker_args
   docker_args:
       value: -p 80:80
```

It also plays nicely with allowing the default behavior to be append.

Reply:
Nice idea, @adamjseitz. I like it. I wonder if it will be possible to use both of them:

```
image: debian:buster
docker_args: --memory 2G

aliases:
  cmd1:
    # This alias ignores the global docker_args
    docker_args: !override !from_yaml some.yml in.here
```

Reply:
I took a first pass at implementing this. 

The implementation is a bit less clean than I would have liked, but it is complicated by the fact that `!override` is treated as a YAML tag, which is parsed first, and then scuba's key-specific `docker_args` parsing occurs later.

This mixin-class based solution allows the object produced by the YAML node to be still accessible with no special handling, but to allow detecting that an `!override` tag has been applied to that object. It does require some special handling around `None`, since it is not desirable to retype the `None` object.

> I wonder if it will be possible to use both of them:

It turns out it is, but with a caveat. The [YAML spec](https://yaml.org/spec/1.2/spec.html#id2788859) does not allow a the content of a plain-style scalar (regardless of the presense of a tag) to begin with a `!`. Thus, the combination is achieved with this syntax:

```yaml
image: debian:buster
docker_args: --memory 2G

aliases:
  cmd1:
    # This alias ignores the global docker_args
    docker_args: !override '!from_yaml some.yml in.here'
```

A potentially more useful application of this is to override with an explicit `!!null` value:

```yaml
image: debian:buster
docker_args: --memory 2G

aliases:
  cmd1:
    # This alias ignores the global docker_args
    docker_args: !override '!!null'
```